### PR TITLE
Add support for opening a file dialog

### PR DESCRIPTION
**Pull request by hant-hub**
Implementing issue f3d-app#1737. Adds new key binding for the f3d application which allows for opening scenes.

### CODE REVIEW

File: application/F3DStarter.cxx [hunk 1]
```diff
@@ -1636,4 +1639,22 @@ void F3DStarter::AddCommands()
         this->LoadFileGroup(index);
       }
     });
+    interactor.addCommand("open_file",
```
Comment (mwestphal) [inline application/F3DStarter.cxx:1642]: look at `add_files` commnad. Your `open_files` should basically do the same after selecting files in the dialog

File: testing/recordings/TestInteractionFileOpen.log [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+KeyPressEvent -546 836 0 0 1 Control_L 0
+CharEvent -546 836 0 0 1 Control_L 0
+TimerEvent -546 836 0 0 1 Control_L 0
+TimerEvent -546 836 0 0 1 Control_L 0
+TimerEvent -546 836 0 0 1 Control_L 0
+TimerEvent -546 836 0 0 1 Control_L 0
+TimerEvent -546 836 0 0 1 Control_L 0
+KeyPressEvent -546 836 2 111 1 o 0
+CharEvent -546 836 2 111 1 o 0
+KeyReleaseEvent -546 836 2 111 1 o 0
```
Comment (mwestphal) [inline testing/recordings/TestInteractionFileOpen.log:11]: I supose there is no way to actually interacti with the dialog from here ?

File: .cache/clangd/index/tinyfiledialogs.h.E0D9CBCF3A31631E.idx
Comment (mwestphal) [inline .cache/clangd/index/tinyfiledialogs.h.E0D9CBCF3A31631E.idx:1]: please remove this file

File: testing/recordings/TestInteractionFileOpen.log (hunk 2 above)
Comment (hant-hub) [reply testing/recordings/TestInteractionFileOpen.log:11]: Not that I am aware of, the dialog function is blocking and I think just queries the OS. The testability is fairly low with tinyfiledialogs.

File: .cache/clangd/index/tinyfiledialogs.h.E0D9CBCF3A31631E.idx
Comment (hant-hub) [reply .cache/clangd/index/tinyfiledialogs.h.E0D9CBCF3A31631E.idx:1]: done in the latest commit

File: application/F3DStarter.cxx (hunk 1 above)
Comment (hant-hub) [reply application/F3DStarter.cxx:1642]: done, I originally found the engine commands when I was first reading through the project so I was using the wrong api. Should be fixed, filename working as expected.

**Review by mwestphal — COMMENTED** [review]
It looks nice! There are still a bunch of improvements needed but lets wait for @Meakk inputs on making this feature optional or not.

File: application/CMakeLists.txt [hunk 3]
```diff
@@ -27,6 +27,8 @@ endif()
 if(APPLE)
   list(APPEND F3D_SOURCE_FILES "${CMAKE_CURRENT_SOURCE_DIR}/F3DNSDelegate.mm")
 endif()
+#add source for dialogs
+list(APPEND F3D_SOURCE_FILES "${F3D_SOURCE_DIR}/external/tinyfiledialogs/tinyfiledialogs.cxx")
```
Comment (Meakk) [inline application/CMakeLists.txt:31]: Isn't it a C file? Why is the extension `cxx` here?

File: doc/THIRD_PARTY_LICENSES.md [hunk 4]
```diff
@@ -202,3 +202,61 @@ DAMAGES, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
...
 ```
+
+
+### tinyfiledialogs
+``` 
+SPDX-License-Identifier: Zlib
+Copyright (c) 2014 - 2024 Guillaume Vareille http://ysengrin.com
+	 ________________________________________________________________
+	|                                                                |
+	| 100% compatible C C++  ->  You can rename this .c file as .cpp |
```
Comment (Meakk) [inline doc/THIRD_PARTY_LICENSES.md:213]: ok I see this comment. That's probably why you changed in to `.cxx`  
I think our build system should be able to handle .c files just fine

File: external/tinyfiledialogs/tinyfiledialogs.h [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+wchar_t * tinyfd_mbcsTo16(char const * aMbcsString);
+char * tinyfd_mbcsTo8(char const * aMbcsString);
+wchar_t * tinyfd_utf8to16(char const * aUtf8string);
+char * tinyfd_utf16to8(wchar_t const * aUtf16string);
+#endif
+/******************************************************************************************************/
+/******************************************************************************************************/
+/******************************************************************************************************/
+
+/************* 3 funtions for C# (you don't need this in C or C++) : */
```
Comment (Meakk) [inline external/tinyfiledialogs/tinyfiledialogs.h:90]: let's add this file in codespell exceptions (see `.codespellrc` file)

File: application/F3DStarter.cxx [hunk 6]
```diff
@@ -704,6 +706,7 @@ class F3DStarter::F3DInternals
       interactor.addBinding({ mod_t::NONE, "Drop" }, "add_files_or_set_hdri", "Others", std::bind(docString, "Load dropped files, folder or HDRI"));
       interactor.addBinding({ mod_t::CTRL, "Drop" }, "add_files", "Others", std::bind(docString, "Load dropped files or folder"));
       interactor.addBinding({ mod_t::SHIFT, "Drop" }, "set_hdri", "Others", std::bind(docString, "Set HDRI and use it"));
+      interactor.addBinding( { mod_t::CTRL, "O" }, "open_file", "Others", std::bind(docString, "Open File"));
```
Comment (Meakk) [inline application/F3DStarter.cxx:709]: This will add the documentation to the cheatsheet, so related tests will fail. You need to update the baselines.

File: application/F3DStarter.cxx [hunk 7]
```diff
@@ -1636,4 +1639,13 @@ void F3DStarter::AddCommands()
         this->LoadFileGroup(index);
       }
     });
+    interactor.addCommand("open_file", 
+      [this](const std::vector<std::string>& args)
```
Comment (Meakk) [inline application/F3DStarter.cxx:1643]:
```suggestion
      [this](const std::vector<std::string>&)
```

File: application/testing/CMakeLists.txt [hunk 8]
```diff
@@ -803,6 +803,7 @@ f3d_test(NAME TestInteractionZoomToggleOrthographicProjection DATA cow.vtp INTER
 f3d_test(NAME TestInteractionRotateCameraMinus90 DATA f3d.glb INTERACTION)
 f3d_test(NAME TestInteractionRotateCamera90 DATA f3d.glb INTERACTION)
 f3d_test(NAME TestInteractionPanWithShift DATA f3d.glb INTERACTION) #Shift;LeftMouse;MouseMovements
+f3d_test(NAME TestInteractionFileOpen DATA cow.vtp INTERACTION NO_BASELINE)
```
Comment (Meakk) [inline application/testing/CMakeLists.txt:806]: It's great and much needed! But I can see that the test is failing because of timeout.

**Review by Meakk — CHANGES_REQUESTED** [review]

File: application/F3DStarter.cxx (hunk 6 above)
Comment (mwestphal) [reply application/F3DStarter.cxx:709]: you can download the new baselines for the CI results directly

File: application/testing/CMakeLists.txt (hunk 8 above)
Comment (mwestphal) [reply application/testing/CMakeLists.txt:806]: Yeah, Im not sure there is an easy way to fix that apart some logic similar to watch test

File: application/CMakeLists.txt [hunk 9]
```diff
@@ -65,6 +67,7 @@ endif ()
 
 target_include_directories(f3d PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}>)
 
+
```
Comment (mwestphal) [inline application/CMakeLists.txt:70]: not needed change

File: application/F3DStarter.cxx [hunk 10]
```diff
@@ -8,7 +8,7 @@
 #include "F3DNSDelegate.h"
 #include "F3DOptionsTools.h"
 #include "F3DPluginsTools.h"
-#include "F3DSystemTools.h"
+#include "scene.h"
```
Comment (mwestphal) [inline application/F3DStarter.cxx:11]: put this include with other libf3d includes

File: application/F3DStarter.cxx (hunk 6 above)
Comment (mwestphal) [inline application/F3DStarter.cxx:709]: I would put this binding before the Drop binding for better ordering

File: application/F3DStarter.cxx [hunk 11]
```diff
@@ -1636,4 +1639,13 @@ void F3DStarter::AddCommands()
         this->LoadFileGroup(index);
       }
     });
+    interactor.addCommand("open_file",
```
Comment (mwestphal) [inline application/F3DStarter.cxx:1642]: open_file_dialog may be a better name

File: application/F3DStarter.cxx [hunk 12]
```diff
@@ -1636,4 +1639,13 @@ void F3DStarter::AddCommands()
...
       }
     });
+    interactor.addCommand("open_file", 
+      [this](const std::vector<std::string>& args)
+      {
+         char* file = tinyfd_openFileDialog( "Open File", ".", 0, nullptr, nullptr, false);
+         if (file) {
+           int index = this->AddFile(file);
+           if (index > -1) this->LoadFileGroup(index);
+         }
```
Comment (mwestphal) [inline application/F3DStarter.cxx:1649]: we use the following style, always:

```
if ()
{
}
```

Comment (hant-hub) [reply application/F3DStarter.cxx:1653]: Should be fixed now, I didn't notice that initially when checking the code style guide.

File: application/F3DStarter.cxx (hunk 7 above)
Comment (hant-hub) [reply application/F3DStarter.cxx]: This should be fixed now. I've mostly been coding in C so I didn't know you could do this in a lambda.

File: application/F3DStarter.cxx (hunk 10 above)
Comment (hant-hub) [reply application/F3DStarter.cxx]: The scene.h was from the incorrect way of updating the scene, its been removed and I put the old include back since its more of a refactoring change and isn't relevant to this issue

File: application/F3DStarter.cxx (hunk 6 above)
Comment (hant-hub) [reply application/F3DStarter.cxx]: Done, I also removed the extra space for better formatting.

Comment (hant-hub) [reply application/F3DStarter.cxx]: I've been poking around the CI artifacts but I am unsure where to find the new baselines. Also when I check the logs it says the tests pass. I'm unsure if I'm supposed to do something here or if this has already been resolved?

File: application/testing/CMakeLists.txt (hunk 8 above)
Comment (hant-hub) [reply application/testing/CMakeLists.txt:806]: f3d says it will use a couple of tools. On my system I have kdialog installed so the opened file dialog is a process of kdialog. If you could add a script that waits for the test inputs to run, then checks for the kdialog process or equivalent that could check for the process then kill it and allow the rest of the tests to run. I'm not sure whether this is a good way to test it since the process name will be different depending on the system and packages.

File: application/F3DStarter.cxx (hunk 6 above)
Comment (mwestphal) [reply application/F3DStarter.cxx]: CI is busted right now, lets focus on the code itself for now, we will fix the tests before the merge anyway :)

File: application/F3DStarter.cxx (hunk 7 above)
Comment (mwestphal) [reply application/F3DStarter.cxx]: Feel free to resolve discussion you adress :)

File: application/F3DStarter.cxx (hunk 6 above)
Comment (mwestphal) [reply application/F3DStarter.cxx]: The doc string should say "Open file dialog" I think.

File: application/CMakeLists.txt (hunk 3 above)
Comment (Meakk) [reply application/CMakeLists.txt:31]: Alright then

File: application/F3DStarter.cxx [hunk 13]
```diff
@@ -1636,4 +1639,17 @@ void F3DStarter::AddCommands()
...
+      {
+         char* file = tinyfd_openFileDialog( "Open File", ".", 0, nullptr, nullptr, false);
+         if (file) 
+         {
+           int index = this->AddFile(file);
+           if (index > -1) 
+           {
+             this->LoadFileGroup(index);
+           }
+         }
```
Comment (Meakk) [inline application/F3DStarter.cxx:1653]: Here a suggestion. You need to define `CTEST_OPEN_DIALOG_FILE` (see other comment).
Wdyt @mwestphal ?

```suggestion
         std::string filePath;
         const char* testFile = getenv("CTEST_OPEN_DIALOG_FILE");
         if (testFile)
         {
           filePath = testFile;
         }
         else
         {
           char* file = tinyfd_openFileDialog( "Open File", ".", 0, nullptr, nullptr, false);
           if (file) 
           {
             filePath = file;
           }
         }

         if (!filePath.empty())
         {
           int index = this->AddFile(file);
           if (index > -1) 
           {
             this->LoadFileGroup(index);
           }
         }
```

File: application/testing/CMakeLists.txt (hunk 8 above)
Comment (Meakk) [inline application/testing/CMakeLists.txt:806]: See previous suggestion
```suggestion
f3d_test(NAME TestInteractionFileOpen DATA cow.vtp INTERACTION NO_BASELINE ENV CTEST_OPEN_DIALOG_FILE="${F3D_SOURCE_DIR}/testing/data/cow.vtp")
```

File: application/F3DStarter.cxx (hunk 13 above)
Comment (hant-hub) [reply application/F3DStarter.cxx]: I implemented this, however this will not be able to test the functionality of tinyfiledialogs, it is only testing the functionality of AddFile and LoadFileGroup. It won't be able to catch if tinyfiledialogs ever breaks

File: application/testing/CMakeLists.txt (hunk 8 above)
Comment (hant-hub) [reply application/testing/CMakeLists.txt]: implemented

File: application/F3DStarter.cxx (hunk 6 above)
Comment (hant-hub) [reply application/F3DStarter.cxx]: just fixed

File: application/F3DStarter.cxx (hunk 10 above)
Comment (hant-hub) [reply application/F3DStarter.cxx]: Undid change, was based on incorrect usage of internal api.

File: application/F3DStarter.cxx (hunk 13 above)
Comment (mwestphal) [reply application/F3DStarter.cxx]: yeah, but I dont think this is actually testable. Its better then nothing though.

File: .codespellrc [hunk 14]
```diff
@@ -1,3 +1,3 @@
 [codespell]
-skip = ./testing/**/*,./.git/**/*,./**/*.patch,./external/cxxopts/cxxopts.hpp,./external/nlohmann_json/nlohmann/json.hpp
+skip = ./testing/**/*,./.git/**/*,./**/*.patch,./external/cxxopts/cxxopts.hpp,./external/nlohmann_json/nlohmann/json.hpp,./external/tinyfiledialogs/tinyfiledialogs.cxx,./external/tinyfiledialogs.h
```
Comment (Meakk) [inline .codespellrc:2]: typo:
```suggestion
skip = ./testing/**/*,./.git/**/*,./**/*.patch,./external/cxxopts/cxxopts.hpp,./external/nlohmann_json/nlohmann/json.hpp,./external/tinyfiledialogs/tinyfiledialogs.cxx,./external/tinyfiledialogs/tinyfiledialogs.h
```

File: application/F3DStarter.cxx [hunk 15]
```diff
@@ -1636,4 +1640,23 @@ void F3DStarter::AddCommands()
...
     });
+    interactor.addCommand("open_file_dialog", 
+      [this](const std::vector<std::string>&)
+      {
+         const char* file = getenv("CTEST_OPEN_DIALOG_FILE");
+         std::cout << file << std::endl;
+         if (!file)
+         {
+           file = tinyfd_openFileDialog( "Open File", ".", 0, nullptr, nullptr, false);
+         }
```
Comment (Meakk) [inline application/F3DStarter.cxx:1651]: One stupid trick to cover the line would be:
```suggestion
         const char* file = getenv("CTEST_OPEN_DIALOG_FILE") || tinyfd_openFileDialog( "Open File", ".", 0, nullptr, nullptr, false);
```

**Review by Meakk — CHANGES_REQUESTED** [review]

File: application/F3DStarter.cxx [hunk 16]
```diff
@@ -698,6 +701,7 @@ class F3DStarter::F3DInternals
       interactor.addBinding({ mod_t::NONE, "Down" }, "add_current_directories", "Others", std::bind(docString, "Add files from dir of current file"));
       interactor.addBinding({ mod_t::NONE, "F11" }, "take_minimal_screenshot", "Others", std::bind(docString, "Take a minimal screenshot"));
       interactor.addBinding({ mod_t::NONE, "F12" }, "take_screenshot", "Others", std::bind(docString, "Take a screenshot"));
+      interactor.addBinding({ mod_t::CTRL, "O" }, "open_file_dialog", "Others", std::bind(docString, "Open File Dialog"));
```
Comment (Meakk) [inline application/F3DStarter.cxx:704]: @mwestphal I'm surprised this change do not impact the cheatsheet tests, is it expected?

Comment (mwestphal) [reply application/F3DStarter.cxx:704]: I would definitely expect this change to impact the baselines

Comment (mwestphal) [reply application/F3DStarter.cxx:704]: How does this tests looks like locally @hant-hub ?

Comment (hant-hub) [reply application/F3DStarter.cxx:704]: All the cheatsheet tests pass. I think its because most of them don't have the part of the cheatsheet visible with the changes, and the one which does only has a small difference. So regardless its 90%+ similar to the baseline.

File: application/F3DStarter.cxx (hunk 15 above)
Comment (hant-hub) [reply application/F3DStarter.cxx:1651]: That would work, but it might be better to have the openFiledialog obviously untested so that in the future its obvious what would and wouldn't be caught.

File: application/F3DStarter.cxx (hunk 16 above)
Comment (mwestphal) [reply application/F3DStarter.cxx:704]: Ill check locally

Comment (mwestphal) [reply application/F3DStarter.cxx:704]: So you were almost right. Most of the cheatsheet tests indeed do not show enough of the cheatsheet to be impacted by this change which is quite low in the cheatsheet. The only one that could have been impacted are the `CheatSheetConfigFile` test since the resolution is made to grab the whole cheatsheet, however it so happens that these tests redefine the `Ctrl+O` interaction effectevely making the interaction added here disapear, which is expected. So all good here :)

File: application/F3DStarter.cxx (hunk 15 above)
Comment (mwestphal) [reply application/F3DStarter.cxx:1651]: I agree with @hant-hub .

Coverage is just a metric, no need to hide that certain part is not tested.

File: application/CMakeLists.txt [hunk 17]
```diff
@@ -27,6 +27,8 @@ endif()
 if(APPLE)
   list(APPEND F3D_SOURCE_FILES "${CMAKE_CURRENT_SOURCE_DIR}/F3DNSDelegate.mm")
 endif()
+#add source for dialogs
```
Comment (mwestphal) [inline application/CMakeLists.txt:30]: add an empty line above

File: application/F3DStarter.cxx [hunk 18]
```diff
@@ -1636,4 +1640,23 @@ void F3DStarter::AddCommands()
         this->LoadFileGroup(index);
       }
     });
+    interactor.addCommand("open_file_dialog", 
+      [this](const std::vector<std::string>&)
+      {
+         const char* file = getenv("CTEST_OPEN_DIALOG_FILE");
```
Comment (mwestphal) [inline application/F3DStarter.cxx:1646]: std::getenv would be better

File: application/F3DStarter.cxx [hunk 19]
```diff
@@ -1636,4 +1640,23 @@ void F3DStarter::AddCommands()
...
+         std::cout << file << std::endl;
```
Comment (mwestphal) [inline application/F3DStarter.cxx:1647]: please remove this log

**Comment by mwestphal** [comment]
Formatting issue to fix:

https://github.com/f3d-app/f3d/actions/runs/12186358862/job/33994681325?pr=1751

**Comment by mwestphal** [comment]
Lets add a .clang-tidy in tinyfiledialogs folder so to exclude all check for these files

File: application/F3DStarter.cxx (hunk 15 above)
Comment (hant-hub) [reply application/F3DStarter.cxx]: I'll resolve this change, we can always add more testing if necessary

File: application/F3DStarter.cxx (hunk 19 above)
Comment (hant-hub) [reply application/F3DStarter.cxx]: removed in latest commit

File: application/F3DStarter.cxx (hunk 18 above)
Comment (hant-hub) [reply application/F3DStarter.cxx]: fixed

File: application/CMakeLists.txt (hunk 17 above)
Comment (hant-hub) [reply application/CMakeLists.txt:31]: should be fixed

File: external/tinyfiledialogs/.clang-tidy
Comment (mwestphal) [inline external/tinyfiledialogs/.clang-tidy:1]: hum, that is not working as I would have expected: 

https://github.com/f3d-app/f3d/actions/runs/12189522943/job/34004897794?pr=1751

Comment (mwestphal) [reply external/tinyfiledialogs/.clang-tidy:1]: Looks like you have to enable at least one check: 

https://stackoverflow.com/questions/58338202/cmake-clang-tidy-disable-checking-in-directory

**Comment by hant-hub** [comment]
I think the only two failing tests now are the code coverage because of the addition of tinyfiledialogs, and the static-analysis possibly also because of tinyfiledialogs. I'm not sure exactly how to either fix them or exempt tinyfiledialogs.

**Review by Meakk — APPROVED** [review]
I'm good with the implementation, thanks @hant-hub !  
As long as the CI is green, let's merge that

File: application/F3DStarter.cxx [hunk 20]
```diff
@@ -9,6 +9,7 @@
 #include "F3DOptionsTools.h"
 #include "F3DPluginsTools.h"
 #include "F3DSystemTools.h"
+#include <cstdlib>
```
Comment (mwestphal) [inline application/F3DStarter.cxx:12]: should go with other std includes

Comment (hant-hub) [reply application/F3DStarter.cxx]: fixed. Accidentally added an include that isn't required

File: external/tinyfiledialogs/.clang-tidy
Comment (hant-hub) [reply external/tinyfiledialogs/.clang-tidy:1]: turns out I forgot to edit the cppcheck suppression file so that was the main issue

**Comment by hant-hub** [comment]
Should be ready to merge.

File: .clang-tidy [hunk 21]
```diff
@@ -2,3 +2,4 @@ Checks: '-*,performance-*,modernize-*,-modernize-use-trailing-return-type,-moder
 WarningsAsErrors: '*'
 UseColor: true
 FormatStyle: none
+
```
Comment (mwestphal) [inline .clang-tidy:5]: is this change needed ? Just curious :)

**Review by mwestphal — APPROVED** [review]
Assuming CI passes :)

**Comment by mwestphal** [comment]
CI is looking good! Merging.
